### PR TITLE
add support for emacs-like key-sequences

### PR DESCRIPTION
**Pull request by ChaosEternal**
added new file: key-names.scm to help translate emacs-style key-names to termbox key definitions

added new proc bind-key and %bind-key to zk.scm,
redefined key binding using bind-key

### CODE REVIEW

**Comment by ChaosEternal** [comment]
add more keynames and replace use of %bind-key with bind-key

File: zk.scm [hunk 1]
```diff
@@ -286,31 +287,81 @@
...
-(hashtable-set! key-table   TB-KEY-ARROW-LEFT	 (lambda () (cursor-left!)))
-(hashtable-set! key-table   TB-KEY-ARROW-RIGHT	 (lambda () (cursor-right!)))
-(hashtable-set! key-table   TB-KEY-CTRL-P	 (lambda () (cursor-up!)))
-(hashtable-set! key-table   TB-KEY-CTRL-N	 (lambda () (cursor-down!)))
-(hashtable-set! key-table   TB-KEY-CTRL-B	 (lambda () (cursor-left!)))
-(hashtable-set! key-table   TB-KEY-CTRL-F	 (lambda () (cursor-right!)))
+(define %bind-key #f)
+(define %list-keys #f)
+(define %match-key #f)
+(let ((key-table (make-eq-hashtable)))
```
Comment (amirouche) [inline zk.scm]: `%key-table` is defined how is this different from `key-table`

File: zk.scm [hunk 2]
```diff
@@ -286,31 +287,81 @@
...
+	      (rest-key (cdr key-seq))
+	      (kt key-table))
+       (if (pair? rest-key)
+	   (let ((cte (hashtable-ref kt ck void)))
+	     (begin (if (not (hashtable? cte))
+			(begin (set! cte (make-eq-hashtable))
+			       (hashtable-set! kt ck cte)))
+		    (lp (car rest-key) (cdr rest-key) cte)))
+	   (hashtable-set! kt ck proc)))))
+  (set! %list-keys
```
Comment (amirouche) [inline zk.scm]: Avoid `set!` when possible.

File: zk/key-names.scm [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      ("C-5" . ,TB-KEY-CTRL-5)
+      ("C-]" . ,TB-KEY-CTRL-RSQ-BRACKET)
+      ("C-RSQ-BRACKET" . ,TB-KEY-CTRL-RSQ-BRACKET)
+      ("C-6" . ,TB-KEY-CTRL-6)
+      ("C-7" . ,TB-KEY-CTRL-7)
+      ("C-/" . ,TB-KEY-CTRL-SLASH)
+      ("C-SLASH" . ,TB-KEY-CTRL-SLASH)
+      ("C-_" . ,TB-KEY-CTRL-UNDERSCORE)
+      ("C-UNDERSCORE" . ,TB-KEY-CTRL-UNDERSCORE)
+      ("C-8" . ,TB-KEY-CTRL-8)))
```
Comment (amirouche) [inline zk/key-names.scm]: There should be a way to do that via module introspection.

File: zk/key-names.scm [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      (cons TB-KEY-ESC (%key-name->keys
+			       (string-append
+				"C-"
+				(substring key-name 4 (string-length key-name))))))
+     ((string=? (substring key-name 0 2) "M-")
+      (cons TB-KEY-ESC (%key-name->keys
+			       (substring key-name 2 (string-length key-name)))))
+     (else '())))
+  (define (%split1 str)
+    (let* ((len (string-length str))
```
Comment (amirouche) [inline zk/key-names.scm]: `let*` is not required here.

File: zk/key-names.scm [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	   (fs (let lp ((i 0))
```
Comment (amirouche) [inline zk/key-names.scm]: I prefer `loop` and `loop0`, `loop1` etc... instead of `lp`...

File: zk.scm [hunk 6]
```diff
@@ -286,31 +287,81 @@
...
+	kt
+	(lambda (k v)
+	  (display `(,prefix ,k ,v) %error)
+	  (newline %error)
+	  (if (hashtable? v)
+	      (%list-keys v (format #f "~a ~a" prefix k )))))]))
+  (set! %dispatch-key
+    (case-lambda
+      [(k) (%dispatch-key k key-table)]
+      [(k kt)
```
Comment (amirouche) [inline zk.scm]: Avoid short hands words like `kt` I prefer to override the previous bindings or better use a `*` suffix like in this case `key-table*`

File: zk.scm [hunk 7]
```diff
@@ -286,31 +287,81 @@
...
+			       (hashtable-set! kt ck cte)))
+		    (lp (car rest-key) (cdr rest-key) cte)))
+	   (hashtable-set! kt ck proc)))))
+  (set! %list-keys
+    (case-lambda
+      [() (%list-keys key-table "")]
+      [(kt prefix)
+       (hash-table-for-each
+	kt
+	(lambda (k v)
```
Comment (amirouche) [inline zk.scm]: avoid multiline `lambda`s

**Comment by amirouche** [comment]
I will merge the changes but I will change a few things to summarize:

- Avoid shorthand words like `lp` `kt` etc..
- Avoid multiline lambdas

And try to document at least public interface.

Thanks!

**Pull request merged**